File: src/main/java/work/lclpnet/ap2/base/activity/MiniGameActivity.java
package work.lclpnet.ap2.base.activity;

import net.minecraft.server.network.ServerPlayerEntity;
import work.lclpnet.activity.ComponentActivity;
import work.lclpnet.activity.component.ComponentBundle;
import work.lclpnet.activity.component.builtin.BossBarComponent;
import work.lclpnet.activity.component.builtin.BuiltinComponents;
import work.lclpnet.ap2.api.base.ParticipantListener;
import work.lclpnet.ap2.api.base.PlayerManager;
import work.lclpnet.ap2.api.game.MiniGame;
import work.lclpnet.ap2.api.game.MiniGameInstance;
import work.lclpnet.ap2.base.cmd.DrawCommand;
import work.lclpnet.ap2.base.cmd.RemakeCommand;
import work.lclpnet.ap2.base.cmd.WinCommand;
import work.lclpnet.ap2.impl.activity.ArcadePartyComponents;
import work.lclpnet.ap2.impl.activity.ScoreboardComponent;
import work.lclpnet.ap2.impl.game.DefaultMiniGameHandle;
import work.lclpnet.ap2.impl.util.scoreboard.CustomScoreboardManager;
import work.lclpnet.kibu.cmd.type.CommandRegistrar;
import work.lclpnet.kibu.hook.HookRegistrar;
import work.lclpnet.kibu.hook.entity.EntityUsePortalCallback;
import work.lclpnet.kibu.hook.player.PlayerAdvancementPacketCallback;
import work.lclpnet.kibu.hook.player.PlayerConnectionHooks;
import work.lclpnet.kibu.hook.player.PlayerRecipePacketCallback;

import java.util.concurrent.atomic.AtomicBoolean;

public class MiniGameActivity extends ComponentActivity {

    private final MiniGame miniGame;
    private final PreparationActivity.Args args;
    private DefaultMiniGameHandle handle;

    public MiniGameActivity(MiniGame miniGame, PreparationActivity.Args args) {
        super(args.container().server(), args.container().logger());
        this.miniGame = miniGame;
        this.args = args;
    }

    @Override
    protected void registerComponents(ComponentBundle componentBundle) {
        componentBundle
                .add(BuiltinComponents.BOSS_BAR)
                .add(BuiltinComponents.COMMANDS)
                .add(BuiltinComponents.HOOKS)
                .add(ArcadePartyComponents.SCORE_BOARD);
    }

    @Override
    public void start() {
        super.start();

        BossBarComponent bossBars = component(BuiltinComponents.BOSS_BAR);

        ScoreboardComponent scoreboardComponent = component(ArcadePartyComponents.SCORE_BOARD);
        CustomScoreboardManager scoreboard = scoreboardComponent.scoreboardManager(args.container()::translations);

        AtomicBoolean remake = new AtomicBoolean(false);

        handle = new DefaultMiniGameHandle(miniGame, args, bossBars, bossBars, scoreboard, remake);
        handle.init();  // hook stack is pushed and later popped by handle::unload in stop()

        PlayerManager playerManager = args.playerManager();

        playerManager.startMiniGame();
        registerHooks(args.container().hookStack());

        MiniGameInstance instance = miniGame.createInstance(handle);
        instance.start();

        ParticipantListener listener = instance.getParticipantListener();
        playerManager.bind(listener);

        CommandRegistrar commands = component(BuiltinComponents.COMMANDS).commands();

        new WinCommand(handle, instance).register(commands);
        new DrawCommand(handle, instance).register(commands);
        new RemakeCommand(handle, remake).register(commands);

        HookRegistrar hooks = component(BuiltinComponents.HOOKS).hooks();
        hooks.registerHook(PlayerAdvancementPacketCallback.HOOK, (player, packet) -> true);
        hooks.registerHook(PlayerRecipePacketCallback.HOOK, (player, packet) -> true);
        hooks.registerHook(EntityUsePortalCallback.HOOK, (entity, portal, pos) -> true);
    }

    @Override
    public void stop() {
        if (handle != null) {
            handle.unload();
        }

        args.playerManager().bind(null);

        handle.getWorldBorderManager().resetWorldBorder();

        super.stop();
    }

    private void registerHooks(HookRegistrar registrar) {
        registrar.registerHook(PlayerConnectionHooks.JOIN, this::onJoin);
        registrar.registerHook(PlayerConnectionHooks.QUIT, this::onQuit);
    }

    private void onJoin(ServerPlayerEntity player) {
        args.container().playerUtil().resetPlayer(player);
    }

    private void onQuit(ServerPlayerEntity player) {
        args.playerManager().remove(player);
    }
}

File: src/modes/default/src/main/kotlin/work/lclpnet/ap2/mode_default/activity/MiniGameActivity.kt
package work.lclpnet.ap2.mode_default.activity

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import net.minecraft.server.level.ServerPlayer
import work.lclpnet.activity.ComponentActivity
import work.lclpnet.activity.component.ComponentBundle
import work.lclpnet.activity.component.builtin.BuiltinComponents
import work.lclpnet.ap2.ext.inWholeTicks
import work.lclpnet.ap2.game.MiniGame
import work.lclpnet.ap2.game.MiniGameInstance
import work.lclpnet.ap2.impl.activity.ArcadePartyComponents
import work.lclpnet.ap2.mode_default.cmd.DrawCommand
import work.lclpnet.ap2.mode_default.cmd.RemakeCommand
import work.lclpnet.ap2.mode_default.cmd.WinCommand
import work.lclpnet.ap2.mode_default.util.ApBaseArgs
import work.lclpnet.ap2.mode_default.util.DefaultMiniGameHandle
import work.lclpnet.ap2.util.MinecraftDispatcher
import work.lclpnet.kibu.hook.HookRegistrar
import work.lclpnet.kibu.hook.entity.EntityUsePortalCallback
import work.lclpnet.kibu.hook.player.PlayerAdvancementPacketCallback
import work.lclpnet.kibu.hook.player.PlayerConnectionHooks
import work.lclpnet.kibu.hook.player.PlayerRecipeNotificationCallback
import java.util.concurrent.atomic.AtomicBoolean

class MiniGameActivity(
    private val miniGame: MiniGame,
    private val args: ApBaseArgs,
) : ComponentActivity(args.miniGameArgs.server, args.miniGameArgs.logger) {

    private lateinit var handle: DefaultMiniGameHandle
    private val scope = CoroutineScope(MinecraftDispatcher(args.miniGameArgs.server) + SupervisorJob())

    override fun registerComponents(componentBundle: ComponentBundle) {
        componentBundle
            .add(BuiltinComponents.BOSS_BAR)
            .add(BuiltinComponents.COMMANDS)
            .add(BuiltinComponents.HOOKS)
            .add(BuiltinComponents.SCHEDULER)
            .add(ArcadePartyComponents.SCORE_BOARD)
    }

    override fun start() {
        super.start()

        val bossBars = component(BuiltinComponents.BOSS_BAR)

        val scoreboardComponent = component(ArcadePartyComponents.SCORE_BOARD)
        val scoreboard = scoreboardComponent.scoreboardManager(args.miniGameArgs::translations)

        val remake = AtomicBoolean(false)

        handle = DefaultMiniGameHandle(miniGame, args, bossBars, bossBars, scoreboard, remake)
        handle.init()  // hook stack is pushed and later popped by handle::unload in stop()

        args.playerManager.startMiniGame()
        registerHooks(args.miniGameArgs.hookStack)

        val hooks = component(BuiltinComponents.HOOKS).hooks()

        PlayerAdvancementPacketCallback.HOOK.registerWith(hooks) { _, _ -> true }
        PlayerRecipeNotificationCallback.HOOK.registerWith(hooks) { _, _, _ -> true }
        EntityUsePortalCallback.HOOK.registerWith(hooks) { _, _, _ -> true }

        val factory = miniGame.createFactory()

        scope.launch {
            val instance = try {
                // suspends until the map is opened, then resumes on the server thread
                factory.createInstance(handle)
            } catch (t: Throwable) {
                args.miniGameArgs.logger.error("Failed to start mini-game {}", miniGame.id, t)
                return@launch
            }

            onInstanceReady(instance, remake)
        }
    }

    private fun onInstanceReady(instance: MiniGameInstance, remake: AtomicBoolean) {
        val listener = instance.participantListener
        args.playerManager.bind(listener)

        val commands = component(BuiltinComponents.COMMANDS).commands()

        WinCommand(handle, instance).register(commands)
        DrawCommand(handle, instance).register(commands)
        RemakeCommand(handle, remake).register(commands)

        val scheduler = component(BuiltinComponents.SCHEDULER).scheduler()
        val maxDurationTicks = instance.maxDuration

        if (maxDurationTicks.isPositive()) {
            scheduler.timeout(maxDurationTicks.inWholeTicks) { ->
                DrawCommand.dispatchDraw(instance)
            }
        }

        args.tablistManager.status = args.miniGameArgs.translations.translateText(miniGame.titleKey)
        args.tablistManager.update()

        instance.start()
    }

    override fun stop() {
        scope.cancel()

        handle.unload()

        args.playerManager.bind(null)

        super.stop()
    }

    private fun registerHooks(hooks: HookRegistrar) {
        PlayerConnectionHooks.JOIN.registerWith(hooks, this::onJoin)
        PlayerConnectionHooks.QUIT.registerWith(hooks, this::onQuit)
    }

    private fun onJoin(player: ServerPlayer) {
        args.miniGameArgs.playerUtil.resetPlayer(player)
    }

    private fun onQuit(player: ServerPlayer) {
        args.playerManager.remove(player)
    }
}